### ISSUE TITLE
systemd stub boot breaks Linux' Secure Boot and Trusted Boot (tboot) detection

### ISSUE THREAD

**Issue by c0d3z3r0** [systemd/systemd#11717]
Debian testing with systemd v239 and v240, on a Lenovo ThinkStation P320
Arch Linux with systemd v241 on a Lenovo ThinkPad x260

I use combined (objcopy), signed kernel images with systemd's linuxx64 stub for direct EFI booting (no bootloader but just BOOTX64.EFI) with Secure Boot enabled on two different machines.
This seems to break Linux' Secure boot detection mechanism (reading boot_params or falling back to UEFI variables if unset):

~~~
[    0.003096] Secure boot could not be determined
~~~

For a quick test I just signed a plain kernel image without initrd or systemd's stub. That works as expected:

~~~
EFI stub: UEFI Secure Boot is enabled.
...
[    0.003142] Secure boot enabled
...
~~~

**Comment by poettering** [comment]
huh, how does the kernel detect that?

**Comment by c0d3z3r0** [comment]
@poettering by reading boot_params supplied by a bootloader or, as fallback, by reading SecureBoot UEFI variable. See https://github.com/torvalds/linux/blob/master/arch/x86/boot/compressed/eboot.c#L786 and https://github.com/torvalds/linux/blob/master/arch/x86/kernel/setup.c#L1163

**Comment by c0d3z3r0** [comment]
The same problem was there in grub: https://bugzilla.redhat.com/show_bug.cgi?id=1418360

**Comment by c0d3z3r0** [comment]
#7482

**Comment by poettering** [comment]
seems somebody miscalculated the offset there... if you correct the offset, does it work for you?

**Comment by c0d3z3r0** [comment]
Which offset do you mean exactly?

Oh, well, I just had a brainstorm.... now I understand what the author did there with these structs...
boot_sector shall be a kind of padding... (size 0x1f1 while setup_params begin at 0x1f1). While this is actually not wrong, not a good coding style... I will correct this, too, when I found a solution for the original issue.

I copied over and adapted (more "hacked together") the structs from the kernel but the problem remains. There must be something other being wrong.

I found this [LWN article](https://lwn.net/Articles/632528/), which is a bit outdated but a bit more detailed than [Documentation/x86/boot.txt](https://www.kernel.org/doc/Documentation/x86/boot.txt).
According to this the bootloader should call `void handover`. This function does not exist in kernel but I think this is `efi_main`, too, but that does not get called, as stated in my last post... I tested this by adding a printk to `efi_main`. Why the heck does the kernel boot at all, then? o.O

**Comment by c0d3z3r0** [comment]
Update: `efi_main` gets called just right... this was my mistake when debugging... I was now able to track this down to linux itself. Everything in systemd's boot stub gets set just fine. 
I will leave this issue open until I found the real cause of this problem.

**Comment by c0d3z3r0** [comment]
I found the issue. After all, it was systemd's implementation of the efi handover protocol. PR is on it's way.

**Comment by poettering** [comment]
Fixed by #11772. CLosing.
